Add a user guide (#140)
strata-org/jverify: docs/guide/src/getting_started.md

# Getting Started

> To get started with JVerify, first follow the directions in the [Installation manual](https://github.com/verus-lang/verus/blob/main/INSTALL.md) to build it.

Reviewer: ```suggestion
To get started with JVerify, first follow the directions in the [Installation manual](https://github.com/aws/jverify/blob/main/INSTALL.md) to build it.
```

Although I'm tickled by "to get started, first give up on Java and use Rust instead" :D

Let's try running JVerify on the following Java program,
found [here](https://github.com/aws/jverify/blob/main/examples/src/test/java/com/aws/jverify/examples/BinarySearch.java):

```java
{{#include ../../../examples/src/test/java/com/aws/jverify/examples/BinarySearch.java}}
```

To run JVerify, run the following from the repository root:

```

> ./verifier/build/install/verifier/bin/verifier ./examples/src/test/java/com/aws/jverify/examples/BinarySearch.java --dafny ./dafny/Scripts/dafny

Reviewer: It would be really nice not to have to specify `--dafny` unless you're using a copy other than the submodule, could we default `--dafny` to `./dafny/Scripts/dafny`?

Reviewer: Or perhaps to just `dafny`, assuming it's on the path somewhere? That will also make it easier to support `dotnet tool` installs of Dafny.

Author: > could we default

But that once works of the working directly is the repository root. Only letting it work automatically if that is the case, seems like it might cause more confusion than it helps.

I've mentioned the option to configure the environment variable `JVERIFY_DAFNY`

> Or perhaps to just dafny, assuming it's on the path somewhere?

It does default to that if nothing else is specified.

Reviewer: > It does default to that if nothing else is specified.

Ah great! Agreed it's better to keep the explicit argument here though, since it's simpler to not require users to mess with their PATH as long as we have the submodule.

# Installation
JVerify can currently only be used by building it from source.

> 1. Ensure the dotnet runtime version 8 or higher is installed on your machine. More information can be found [here](https://dotnet.microsoft.com/en-us/).

Reviewer: ```suggestion
1. Ensure the .NET runtime version 8 or higher is installed on your machine. More information can be found [here](https://dotnet.microsoft.com/en-us/).
```

> 1. Check out the JVerify repository and its submodules, for example using `git clone https://github.com/aws/jverify --recurse-submodules`.

Reviewer: ```suggestion
1. Check out the [JVerify repository](https://github.com/aws/jverify) and its submodules, for example using `git clone https://github.com/aws/jverify --recurse-submodules`.
```

1. Navigate to the JVerify repository directory.
1. Run `./gradlew installDist` (for Mac OS or Linux systems) or `./gradlew.bat installDist` (for Windows)

> 1. Run `make -C dafny exe` to build [Dafny](https://dafny.org/), a tool used by JVerify.

Reviewer: ```suggestion
1. Run `make -C dafny exe` to build [Dafny](https://dafny.org/), a verification tool used by JVerify.
```

1. Install Z3 using `make -C dafny z3-<OS>`, where `<OS>` can be `mac`, `mac-arm` or `ubuntu`. Pick the one that matches your OS. For systems that do not fit those categories, such as Windows, you will have to take these steps instead:
    1. Download the release of [Z3 version 4.12.1](https://github.com/Z3Prover/z3/releases/tag/z3-4.12.1) that matches your system.

>     1. Unzip the content of the Z3 folder into dafny\Binaries\z3 so that dafny\Binaries\z3\bin\z3.exe exists.

Reviewer: ```suggestion
    1. Unzip the content of the Z3 folder into `dafny\Binaries\z3` so that `dafny\Binaries\z3\bin\z3.exe` exists.
```

# Usage

Let's try running JVerify on the following Java program,
found [here](https://github.com/aws/jverify/blob/main/examples/src/test/java/com/aws/jverify/examples/BinarySearch.java):

```java
{{#include ../../../examples/src/test/java/com/aws/jverify/examples/BinarySearch.java}}
```

To run JVerify, run the following from the repository root:

```
./verifier/build/install/verifier/bin/verifier ./examples/src/test/java/com/aws/jverify/examples/BinarySearch.java --dafny ./dafny/Scripts/dafny
```

or the following in Windows:

```
./verifier/build/install/verifier/bin/verifier.bat ./examples/src/test/java/com/aws/jverify/examples/BinarySearch.java --dafny ./dafny/Scripts/dafny
```

You should see the following output:

```
Dafny program verifier finished with 3 verified, 0 errors
```

> This indicates that JVerify successfully verified 3 parts.

Reviewer: ```suggestion
This indicates that JVerify successfully verified 3 batches of specifications.
```

(Hard to do much better before #127 though)

> You can remove the need for the `--dafny` option by configuring the environment variable `JVERIFY_DAFNY` instead.

Reviewer: ```suggestion
You can remove the need for the `--dafny` option by configuring the environment variable `JVERIFY_DAFNY` instead, or modifying your `PATH` so that `dafny` resolves automatically.
```

?
